flink-connectors/flink-connector-upsert-test/src/main/java/org/apache/flink/connector/upserttest/sink/BinaryFileUtil.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.upserttest.sink;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.serialization.DeserializationSchema;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Collection uf utility methods for tests using the {@link UpsertTestSink}. */
+@Internal
+public class BinaryFileUtil {
+    private static final byte MAGIC_BYTE = 13;
+
+    /**
+     * Reads records that were written using the {@link BinaryFileUtil#readRecords} method from the
+     * given InputStream.
+     *
+     * @param bis The BufferedInputStream to read from
+     * @return Map containing the read ByteBuffer key-value pairs
+     * @throws IOException
+     */
+    public static Map<ByteBuffer, ByteBuffer> readRecords(BufferedInputStream bis)
+            throws IOException {
+        checkNotNull(bis);
+        Map<ByteBuffer, ByteBuffer> records = new HashMap<>();
```

Comment:
I am afraid using `ByteBuffers` as a key does not work. I looked at the code of `ByteBuffer#hashCode` and the implementation is completely independent of the actual content of the buffer. The hash is calculated from the contained metadata information that easily might lead to unwanted collisions.

Reply:
Where do you see this? The one I found is iterates through the contents individual bytes:
```
public int hashCode() {
        int h = 1;
        int p = position();
        for (int i = limit() - 1; i >= p; i--)
            h = 31 * h + (int)get(i);
        return h;
    }
```